<a href="https://colab.research.google.com/github/j-claudinei-f/j-claudinei-f/blob/main/Inducao_e_Recorrencia.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#**Texto auxiliar sobre demonstrações por indução e relações de recorrência**

José Claudinei Ferreira

Universidade Federal de Alfenas

#**Uma infinidade de números primos**



Um número natural positivo que não pode ser escrito como produto de dois outros números naturais menores que ele é chamado de número primo.

**Teorema:** Existem infinitos números primos.

**Demonstração:** Sabemos que um número natural $n$ é primo quando ele é divisível apenas por 1 e por ele mesmo.

Suponha que o teorema seja falso. Isso implicaria que o conjunto $P$ de todos os números primos seria finito, ou seja, teria $p$ elementos, sendo $p$ um número natural.

Então, poderíamos enumerar esses elementos de $P$ como $$P=\{n_1,\,n_2,\,\ldots,\, n_p\}.$$

Se esse fosse o caso, o número natural $$n_{p+1}=n_1\times n_2\times\cdots\times n_p+1$$ <font color=blue> não seria divisivel por nenhum $n_j\in P$ e, portanto, seria primo.

Isso geraria um problema, porque supomos que existiam apenas $p$ números primos e não $p+1$.

Logo, $P$ não pode ser finito.$\hspace{4cm}\square$

**Obs:** Não é sempre verdade que o número $n_{p+1}$ da demonstração é primo. Seria verdade se $P$ fosse finito, o que não é o caso. Mas se você testar alguns casos pode encontrar primos.

In [ ]:
import sympy as sp # para usar a função isprime() que testa se um número é primo.

Testendo alguns casos:

In [ ]:
p=2*3+1
p,sp.isprime(p)

(7, True)

In [ ]:
p=2*3*5+1
p,sp.isprime(p)

(31, True)

In [ ]:
p=2*3*5*7+1
p,sp.isprime(p)

(211, True)

In [ ]:
p=2*3*5*7*11+1
p,sp.isprime(p)

(2311, True)

In [ ]:
p=2*3*5*7*11*13+1
p,sp.isprime(p)

(30031, False)

**Obs:** <font color=blue> Na verdade, na demonstração anterior apresentada usamos um resultado que podemos provar.

**Teorema 2:**

Um número natural $n$ ou é primo ou é o produto $n=n_1\times n_2$, sendo $n_1$ um número primo.

**Demonstração:** <font color=blue> Feito em sala.

**Corolário:** Um número natural $n$ ou é primo ou é o produto de números primos.

**Demonstração:** <font color=blue> Feito em sala.

**Corolário:** Um número natural $n$ ou é primo ou é o produto  $n=n_1\times n_2$, com $n_1$ primo e $1<n_1\leq \sqrt{n}$.

**Demonstração:** <font color=blue> Feito em sala, mas repito aqui. Do teorema temos que $n=n_1\times n_2$. Se $n_1>\sqrt{n}$ e $n_2>\sqrt{n}$, então $n_1\times n_2>\sqrt{n}\times \sqrt{n}=n$, o que não pode ocorrer.<br><br>
Logo, $n_1$ ou $n_2$ tem que ser menor ou igual a $\sqrt{n}$; e podemos escolher esse número menor primo.$\hspace{2cm}\square$

Esses resultados nos permitem verificar se um número é primo ou determinar seus fatores, como o algoritmo que segue:

In [ ]:
import math

def Fatores_primos(n):     # Só funciona para números naturais maiores que 1.
    while n % 2 == 0:      # imprime o número de vezes que 2 divide n.
        print(2)
        n = n // 2         # troca n por n/2 (divisão inteira)

    i=3
    while i*i<=n:          # testando números ímpares.
         while n % i== 0:  # imprime o número de vezes que i divide n.
            print(i)
            n = n // i     # troca n por n/i (divisão inteira)
         i=i+2

    if n > 2:
        print(n) # Se não hove divisão de $n$ ele é primo

**Obs:** Note que, caso prefira (ou possa) usar o cálculo de $\sqrt{n}$, você pode trocar o

```
while i*i<=n:
```
por

```
for i in range(3,int(math.sqrt(n))+1,2):
```

**Obs:** Uma curiosidade sobre o cálculo de $\sqrt{x}$ no [link](https://en.wikipedia.org/wiki/Fast_inverse_square_root).

Teste com $n=2\times3\times5\times7\times11\times13+1=30031$ que já testamos não ser primo.



In [ ]:
p=2*3*5*7*11*13+1
Fatores_primos(p)

59
509


**Obs:** <font color=blue> Note que $30031=59\times 509$ é produto de dois números primos, um menor e outro maior que sua raiz quadrada.

<font color=red> Isso nos diz que pode ocorrer um fator primo maior que a raiz quadrada, mas só pode ocorrer um.

Vamos testar com outro número:

In [ ]:
Fatores_primos(123456789)

3
3
3607
3803


Quando esse algoritmo retornar apenas o número $n$ de entrada, esse número será primo.

In [ ]:
Fatores_primos(2*3*5*7*11+1)

2311


#**A torre de Hanói**

No estudo da determinação do número mínimo de movimentos para mover os discos da [torre de Hanói](http://clubes.obmep.org.br/blog/torre-de-hanoi/) observamos que, se denotarmos por $p(n)$ o número mínimo de movimentos, para mover $n$ discos, com as **regras**:


1.   Mover um disco de cada vez.
2.   Não colocar disco maior sobre disco menor.

Teremos
$$p(1)=1,\qquad p(n+1)=2p(n)+1.$$

Como podemos determinar $p(6)$ e $p(10)$?





**Uma forma de resolver o problema, com o uso de um computador e programação:**

Podemos usar uma linguagem de programação (ou uma calculadora programável) para calcular $p(n)$. Nesse caso, vamos usar a linguagem Phyton.

In [ ]:
def p(n):         # O comando def é usado para definir funções das mais variadas.

   if (n==1):     # Se n=1, retorne p(1)=1
     q=1

   else:          # Senão, ou seja, se p for diferente de 1, retorne 2p(n-1)+1.
     q=2*p(n-1)+1 # Aqui a função é chamada outra vez, para m=n-1, e outra vez, para m=n-2, ..., até chegar em m=1.
                  # Após todas as chamadas retorna a m=n-1 e m=n.

   return q       # Essa é a saída.

**Obs.** Note que a função não está bem definida para números menores que 1, ou para números que não sejam naturais.

Vamos testar, para $n=1$:

In [ ]:
p(1)

1

Vamos testar, para $n=2$:

In [ ]:
p(2)

3

Vamos testar, para $n=3$:

In [ ]:
p(3)

7

Para ver vários valores de $p(n)$ de uma vez, por exemplo, $p(i)$, para $1\leq i\leq 10$:

In [ ]:
for i in range(1,10+1): # Para i variando de 1 até 10, escreva o par (i, p(i)).
  print(i,p(i))

1 1
2 3
3 7
4 15
5 31
6 63
7 127
8 255
9 511
10 1023


**Obs.** Voce poderia fazer isso de outra forma, sem usar a definição da função $p(i)$. Começando a enumerar todos os $p(i)$, a partir do 1, como fazemos a seguir:

In [ ]:
q=1                   # q=1 representa p(1)=1.
print(1,q)

for i in range(1,10): # Para i variando de 1 até 10, troque q por 2q+1, porque p(n)=2p(n-1)+1.
  q=2*q+1
  print(i+1,q)        # Escreva o par (i, 2p+1).

1 1
2 3
3 7
4 15
5 31
6 63
7 127
8 255
9 511
10 1023


Esse último procedimento é menos elegante, mas é muito mais rápido de ser executado.

**Obs.** Observe ainda que podemos demonstrar, usando [indução finita](https://pt.wikipedia.org/wiki/Indu%C3%A7%C3%A3o_matem%C3%A1tica), que $$p(n)=2^n-1.$$ Isso facilita bastante na resolução do problema, embora uma calculadora ajude muito ainda.

In [ ]:
for i in range(1,10+1): # Para i variando de 1 até 10.
    print(i,2**i-1)        # Escreva o par (i, 2^i-1).

1 1
2 3
3 7
4 15
5 31
6 63
7 127
8 255
9 511
10 1023


**Complicando um pouco mais a regra:**

No estudo da determinação do número mínimo de movimentos para mover os discos da [torre de Hanói](http://clubes.obmep.org.br/blog/torre-de-hanoi/) observamos que, se denotarmos por $p(n)$ o número mínimo de movimentos, para mover $n$ discos, com as **regras**:


1.   Mover um disco de cada vez.
2.   Não colocar disco maior sobre disco menor.
3.   Cada movimento de disco deve partir da haste central ou chegar nela.

Teremos
$$p(1)=2,\qquad p(n+1)=?.$$

Como podemos determinar $p(6)$ e $p(10)$?

#**A regra de Lapace para calcular determinantes**

Há relações de recorrência que não podem ser escritas (de forma simples) por enumeração dos itens anteriores. Um exemplo disso é o cálculo de determinante de uma matriz, por meio da [regra de Laplace](https://pt.wikipedia.org/wiki/Teorema_de_Laplace):
$$det(A)=\sum_{j=1}^n(-1)^{j+1}a_{1j}det(A_{1,j}),$$ em que $A=[a_{ij}]$ é uma matriz $n\times n$ e $A_{i,j}$ é uma matriz de ordem $(n-1)\times (n-1)$, obtida retirando a linha $i$ e a coluna $j$ da matriz $A$. Nessa definição temos $1\leq i,j\leq n$.


In [ ]:
import numpy as np   # Pacote para manipular matrizes.

def Cof(A,i,l):      # Determinação da matriz A_{i,l}.
   n=np.size(A[0])
   C=[]
   for j in range(0,n):
    B=[]
    if j!=i:
      for k in range(0,n):
        if k!=l:
          B.append(A[j][k])
      C.append(B)
   return C

In [ ]:
import pandas as pd   # Pacote para manipular tabelas.

A=np.array([[0,1,-4,7,6],[3,3,1,2,-9],[1,4,2,6,3],[1,5,9,8,7],[3,-2,-4,7,5.0]])
                      # Foi colocado 5.0 no último número para dizer à máquina
                      # que os números são reais e não inteiros.

print('Matriz A=\n',pd.DataFrame(A))
print('\n A matriz menor 1,2 é\n',pd.DataFrame(Cof(A,1,0)))

Matriz A=
      0    1    2    3    4
0  0.0  1.0 -4.0  7.0  6.0
1  3.0  3.0  1.0  2.0 -9.0
2  1.0  4.0  2.0  6.0  3.0
3  1.0  5.0  9.0  8.0  7.0
4  3.0 -2.0 -4.0  7.0  5.0

 A matriz menor 1,2 é
      0    1    2    3
0  1.0 -4.0  7.0  6.0
1  4.0  2.0  6.0  3.0
2  5.0  9.0  8.0  7.0
3 -2.0 -4.0  7.0  5.0


In [ ]:
def detLaplace(A):     # Cálculo de determinante por regra de Laplace.
  n=np.size(A[0])
  if n==1:
    p=A[0][0]
  else:
    p=0
    for i in range(0,n):
      p=p+(-1)**((i+1)+1)*A[0][i]*detLaplace(Cof(A,0,i)) # O índice aqui começa no 0.
  return p

In [ ]:
detLaplace(A)

-3318.0

O cálculo de terminante é muito usado, e por isso tem pacote pronto. Note que o pacote gera arredondamentos.

In [ ]:
np.linalg.det(A)

-3318.000000000002

**Da dificuldade de usar a regra de Laplace**

Podemos dizer que, se $p(n)$ denota o número de cálculos na determinação do determinante de uma matriz $n\times n$, por esta regra de Laplace, então
$$p(n)\geq n\times p(n-1)+3n,\qquad n\geq 2,\qquad p(1)=1,$$ porque precisamos calcular $n$ determinantes, com $p(n-1)$ cálculos cada, e multiplicar esses valores por $(-1)^{1+j}a_{1j}$, o que totaliza $3n$ cálculos (pelo menos).

Então,
$$\begin{cases}p(n)&\geq &np(n-1)+3n\\\\&\geq&n((n-1)p(n-2)+3(n-1))+3n\\\\\vdots&\vdots&\vdots\\\\&\geq& n(n-1)\cdots2p(1)+3(n+(n-1)+\cdots+2)\\\\&=&n!+3\left(\frac{n(n+1)}{2}-1\right)\end{cases}$$

Nesse argumento não fica claro o problema de armazenamento de todos determinantes necessários no cálculo, o que pode levar a problemas de memória por causa de [pilhas](https://algoritmosempython.com.br/cursos/algoritmos-python/estruturas-dados/pilhas/).

**A matriz inversa, se existir:**

Conceitualmente o cálculo de determinante é útil na determinação da [matriz inversa](https://pt.wikipedia.org/wiki/Matriz_inversa) de $A$, dada por
$$A^{-1}=\frac{1}{det(A)}\begin{bmatrix}C_{1,1}&C_{1,2}&\cdots&C_{1,n}\\C_{2,1}&C_{2,2}&\cdots&C_{2,n}\\\vdots&\vdots&\ddots&\vdots\\C_{n,1}&C_{n,2}&\cdots&C_{n,n}\end{bmatrix}^t,$$ em que $C_{i,j}=(-1)^{i+j}det(A_{i,j})$.

Entretanto, esse método de determinar $A^{-1}$ é bastante caro computacionalmente, quando usamos regra de Lapalace. Podemos dizer que o custo computacional $c(n)$ satisfaz $$c(n)\geq n\times p(n-1)+3n+n^2((n-1)p(n-2)+3(n-1)),$$ porque precisamos calcular um determinante de uma matriz $n\times n$ e $n^2$ determinantes de matrizes $(n-1)\times (n-1)$, pelo menos.

In [ ]:
def Mult_inv(A): # Gera uma matriz p*A^(-1), em que p=det(A).
  n=np.size(A[0])
  B=[]
  for i in range(0,n):
    C=[]
    for j in range(0,n):
      C.append((-1)**(i+j+2)*detLaplace(Cof(A,j,i))) # Na expressão tem um transposto.
    B.append(C)
  return np.array(B)


Testando o nosso código, observamos que a matriz $B$ obtida é tal que $B\times A$ tem diagonal igual a $det(A)$.

In [ ]:
Mult_inv(A)@A      # @ denota produto de matrizes. O * faz multiplicação coordenada a coordenada.

array([[-3318.,     0.,     0.,     0.,     0.],
       [    0., -3318.,     0.,     0.,     0.],
       [    0.,     0., -3318.,     0.,     0.],
       [    0.,     0.,     0., -3318.,     0.],
       [    0.,     0.,     0.,     0., -3318.]])

**Problema 1:** Usando a exepressão envolvendo $c(n)$ acima. Determine uma aproximação, por baixo, para $c(10)$, $c(20)$ e $c(100)$.

**O escalonamento é mais barato**

É por isso que o processo de [eliminação de Gauss](https://pt.wikipedia.org/wiki/Elimina%C3%A7%C3%A3o_de_Gauss) é preferido para cálculo de determinantes, por exemplo. Esse processo de eliminação pode ser visto como recorrente, ou seja, a eliminação de uma coluna depende a eliminação da coluna anterior.

In [ ]:
import copy

def busca_pivo(B):
    A=copy.deepcopy(B)
    i=0
    while abs(A[i,0])<10**(-15): # Precisa tocar linha para eliminação.
      i=i+1
    p=copy.deepcopy(A[i])
    A[i]=-copy.deepcopy(A[0])     # Isso é para manter o sinal do determinante.
    A[0]=p
    return A

In [ ]:
print('A=\n',A,'\n\n A com linhas trocadas\n',busca_pivo(A))

A=
 [[ 0.  1. -4.  7.  6.]
 [ 3.  3.  1.  2. -9.]
 [ 1.  4.  2.  6.  3.]
 [ 1.  5.  9.  8.  7.]
 [ 3. -2. -4.  7.  5.]] 

 A com linhas trocadas
 [[ 3.  3.  1.  2. -9.]
 [-0. -1.  4. -7. -6.]
 [ 1.  4.  2.  6.  3.]
 [ 1.  5.  9.  8.  7.]
 [ 3. -2. -4.  7.  5.]]


In [ ]:
def Gauss_Elim(B):
  A=copy.deepcopy(B)                    # Para não alterar a matriz A no processo.
  n=np.size(A[0])
  print(pd.DataFrame(A),'\n')           # Mostra a matriz A
  if n==0:
    p=A
  else:
    A=busca_pivo(A)
    for i in range(1,n):
      A[i]=A[i]-A[0]*A[i,0]/A[0,0]
    print(pd.DataFrame(A),'\n')         # Mostra a matriz A após eliminação da primeira coluna.
    if n>2:
      A[1:n,1:n]=Gauss_Elim(A[1:n,1:n]) # Recorrência
  p=A
  return p

In [ ]:
B=Gauss_Elim(A)
print(pd.DataFrame(B))

     0    1    2    3    4
0  0.0  1.0 -4.0  7.0  6.0
1  3.0  3.0  1.0  2.0 -9.0
2  1.0  4.0  2.0  6.0  3.0
3  1.0  5.0  9.0  8.0  7.0
4  3.0 -2.0 -4.0  7.0  5.0 

     0    1         2         3     4
0  3.0  3.0  1.000000  2.000000  -9.0
1  0.0 -1.0  4.000000 -7.000000  -6.0
2  0.0  3.0  1.666667  5.333333   6.0
3  0.0  4.0  8.666667  7.333333  10.0
4  0.0 -5.0 -5.000000  5.000000  14.0 

     0         1         2     3
0 -1.0  4.000000 -7.000000  -6.0
1  3.0  1.666667  5.333333   6.0
2  4.0  8.666667  7.333333  10.0
3 -5.0 -5.000000  5.000000  14.0 

     0          1          2     3
0 -1.0   4.000000  -7.000000  -6.0
1  0.0  13.666667 -15.666667 -12.0
2  0.0  24.666667 -20.666667 -14.0
3  0.0 -25.000000  40.000000  44.0 

           0          1     2
0  13.666667 -15.666667 -12.0
1  24.666667 -20.666667 -14.0
2 -25.000000  40.000000  44.0 

           0          1          2
0  13.666667 -15.666667 -12.000000
1   0.000000   7.609756   7.658537
2   0.000000  11.341463  22.048780 

Agora multiplicamos os elementos da diagonal da matriz obtida na eliminação, e temos o determinante de $A$.

In [ ]:
n=np.size(B[0])
p=1
for i in range(0,n):
  p=p*B[i,i]
p

-3317.9999999999995

Para a eliminação da primeira coluna de $A$, é necessário fazermos $n-1$ divisões (nos pivos), $(n-1)(n-1)$ produtos (pivo e entrada) e $(n-1)(n-1)$ somas (linha pivo com outras linhas), ou $(n-1)(2n-1)$ operações. Então, a quantidade $g(n)$ de somas e produtos para a eliminação toda será
$$g(n)=g(n-1)+2(n-1)^2+(n-1),\qquad g(2)=3.$$

**Problema 4:** Considerando a expressão para $g(n)$, o último código para o calculo de determinate, por meio de eliminação, e a expressão para $A^{-1}$ apresentada acima.

Determine quantas operações, no mínimo, serão necessárias para calcular $A^{-1}$, sendo $A$ uma matriz $n\times n$ invertível, usando eliminação em vez de regra de Laplace para o cálculo de determinantes:

a) De forma recorrente;

b) De forma direta.